pat/ReferenceTestV2/consumer/step_4.py

```diff
@@ -0,0 +1,436 @@
+"""Tests metric updxates."""
+
+from __future__ import annotations
+
+import collections
+import functools
+import logging
+import os
+import queue
+import threading
+import time
+from typing import TYPE_CHECKING
+
+from pat.ReferenceTestV2.consumer import result_collector
+from sdc11073.observableproperties import observables
+from sdc11073.xml_types import pm_qnames
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from sdc11073 import xml_utils
+    from sdc11073.mdib import ConsumerMdib, statecontainers
+
+
+__STEP__ = '4'
+logger = logging.getLogger(f'pat.consumer.step_{__STEP__}')
+
+
+def _count_states_occurrences(  # noqa: PLR0913
+    step: str,
+    update_queue: queue.Queue,
+    updates_required_count: int,
+    node_types: Iterable[xml_utils.QName],
+    collected_updates_within_timeout_event: threading.Event,
+    states_by_handle: dict[str, statecontainers.AbstractStateContainer],
+):
+    for state in states_by_handle.values():
+        if state.NODETYPE in node_types:
+            logger.debug(
+                'State "%s" of type "%s" received',
+                state.DescriptorHandle,
+                state.NODETYPE,
+                extra={'step': step},
+            )
+            update_queue.put_nowait(state)
+            if update_queue.qsize() >= updates_required_count:
+                collected_updates_within_timeout_event.set()
+
+
+def _verify_state_updates_in_time(
+    mdib: ConsumerMdib,
+    step: str,
+    node_types: Iterable[xml_utils.QName],
+    updates_required_count: int,
+    timeout: float,
+):
+    collected_updates_within_timeout = threading.Event()
+
+    update_queue = queue.Queue()
+    observer = functools.partial(
+        _count_states_occurrences,
+        step,
+        update_queue,
+        updates_required_count,
+        node_types,
+        collected_updates_within_timeout,
+    )
+
+    if any(
+        node_type
+        for node_type in node_types
+        if node_type in (pm_qnames.NumericMetricState, pm_qnames.StringMetricState, pm_qnames.EnumStringMetricState)
+    ):
+        mdib_observer = 'metrics_by_handle'
+    elif any(
+        node_type
+        for node_type in node_types
+        if node_type in (pm_qnames.AlertConditionState, pm_qnames.LimitAlertConditionState, pm_qnames.AlertSignalState)
+    ):
+        mdib_observer = 'alert_by_handle'
+    elif any(node_type for node_type in node_types if node_type in (pm_qnames.RealTimeSampleArrayMetricState,)):
+        mdib_observer = 'waveform_by_handle'
+    elif any(
+        node_type
+        for node_type in node_types
+        if node_type in (pm_qnames.ClockState, pm_qnames.BatteryState, pm_qnames.MdsState, pm_qnames.VmdState)
```

Comment:
I don't know if you want to have this lists final, but e.g. ChannelState is missing here
(also AlertSystem is missing here)